src/operator/batch_norm-inl.h

```diff
@@ -131,70 +133,74 @@ class BatchNormOp : public Operator {
     CHECK_EQ(out_data.size(), 3);
     CHECK_EQ(in_grad.size(), 3);
     Stream<xpu> *s = ctx.get_stream<xpu>();
-    Tensor<xpu, 4> data, grad, grad_in;
-    const real_t scale = static_cast<real_t>(out_grad[batchnorm::kOut].shape_[1]) /
-                         static_cast<real_t>(out_grad[batchnorm::kOut].shape_.Size());
+    Tensor<xpu, 4, DType> data, grad, grad_in;
+    const DType scale = DType(
```

Comment:
Ye, It should be handled like the one in forward. About the FP16 cudnn BN, those mean and variance are typed in float not as half. I'll download these code and update them with a fix. Maybe you can double check them later. There is some issues with mshadow implementation using FP16 according to my experiment. The fp32 one works fine but fp16 one fails.  can you check that one later after I fixes current issues?